Adding R8 deobfuscation to otlp batch processors

LikeTheSalad opened the pull request:
<!-- Thanks for sending a pull request!

If this is your first contribution, please review and sign our contributor agreement -
https://www.elastic.co/contributor-agreement.

Guidelines:
 - Prefer small PRs, and split changes into multiple logical commits where they must
   be delivered in a single PR.
 - If the PR is incomplete and not yet ready for review, open it as a Draft.
 - Once the PR is marked ready for review it is expected to pass all tests and linting,
   and you should not force-push any changes.

See also https://github.com/elastic/apm-server/blob/main/CONTRIBUTING.md for more tips on contributing.
-->

## Motivation/summary

<!--
Describe your change in the title and description, and provide a motivation for the
change and rationale for the approach taken.
-->

This change will allow doing an Android exception stacktrace deobfuscation when receiving an OTLP mobile exception log. In order to do so, part of this PR is related to fetching an Android deobfuscation map file from Elasticsearch, which is currently done in a similar way to how it's done for JS RUM Source maps. In [the first draft](https://github.com/elastic/apm-server/pull/10744), the idea was to reuse the same `Fetcher` interface as the one used for JS Source maps, however, it was [later mentioned](https://github.com/elastic/apm-server/pull/10744#issuecomment-1561735340) that said interface is too tied to the specific JS RUM use case, making it complicated to be reused for other purposes. Because of that reason, and because the overall Android mapping process hasn't been fully defined yet, the way the Android map file is fetched as part of this PR doesn't involve reusing the `Fetcher` interface, but picking only parts of the `Fetcher` implementation related to the Elasticsearch query that would help fetching the Android map file, without having to touch the existing JS RUM interface nor its implementations for now.

Once the first iteration is working, I believe we would get a better overview of the similarities across both types of fetching mechanisms, so that we can pinpoint what parts of them can be shared and do the refactoring accordingly.

## Goals

* Add a processor that deobfuscates Android stacktraces for the OTLP data ingestion.
* Fetch android maps from ES.
* Store maps in cache and reuse them.

## Checklist

<!--
Delete irrelevant items. The changelog should only be updated for user-facing changes.
Once the PR is ready for review there should be no unticked boxes.
-->

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## How to test these changes

<!--
Explain how this PR can be tested by the reviewer: commands, dependencies, steps, etc.
If it is self-explanatory, delete this section.
-->

## Related issues

<!--
Reference the related issue(s), and make use of magic keywords where it makes sense
https://help.github.com/articles/closing-issues-using-keywords/.
-->

LikeTheSalad commented:
I'm planning to wait for the new Kibana endpoint (to upload new Android map files to Elasticsearch) to be available before I can fully finish this PR and mark it as ready for review.

apmmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-11005/detail/PR-11005/2//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-11005/detail/PR-11005/2//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-11005/detail/PR-11005/2//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-11005/detail/PR-11005/2//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-server_11005.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/apm-ci/transactions/view?rangeFrom=2023-07-17T10:40:52.851Z&rangeTo=2023-07-17T11:00:52.851Z&transactionName=BUILD+apm-server%2Fapm-server-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=2ad2ecf99d6d953591bf68ba7b20a0ca&transactionId=774974b876cbe189)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-07-17T10:50:52.851+0000


* Duration: 6 min 2 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate and publish the docker images.


- `/test windows` : Build & tests on Windows.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

apmmachine commented:
### :books: Go benchmark report

Diff with the `main` branch

```
goos: linux
goarch: amd64
pkg: github.com/elastic/apm-server/internal/agentcfg
cpu: 12th Gen Intel(R) Core(TM) i5-12500
                                  │ build/main/bench.out │             bench.out             │
                                  │        sec/op        │   sec/op     vs base              │
geomean                                      67.82n        67.58n       -0.35%

                                  │ build/main/bench.out │             bench.out              │
                                  │         B/op         │    B/op     vs base                │
geomean                                                ²               +0.00%               ²
¹ all samples are equal
² summaries must be >0 to compute geomean

                                  │ build/main/bench.out │             bench.out              │
                                  │      allocs/op       │ allocs/op   vs base                │
geomean                                                ²               +0.00%               ²
¹ all samples are equal
² summaries must be >0 to compute geomean

pkg: github.com/elastic/apm-server/internal/beater/request
                                             │ build/main/bench.out │              bench.out              │
                                             │        sec/op        │    sec/op     vs base               │
ContextReset/X-Real-IP_ipv6-12                         593.6n ± 25%   783.7n ± 42%  +32.02% (p=0.009 n=6)
geomean                                                942.0n         907.4n         -3.67%

                                             │ build/main/bench.out │              bench.out               │
                                             │         B/op         │     B/op      vs base                │
geomean                                                           ²                 +0.00%               ²
¹ all samples are equal
² summaries must be >0 to compute geomean

                                             │ build/main/bench.out │             bench.out              │
                                             │      allocs/op       │ allocs/op   vs base                │
geomean                                                           ²               +0.00%               ²
¹ all samples are equal
² summaries must be >0 to compute geomean

pkg: github.com/elastic/apm-server/internal/publish
             │ build/main/bench.out │          bench.out          │
             │        sec/op        │   sec/op    vs base         │

             │ build/main/bench.out │           bench.out            │
             │         B/op         │     B/op       vs base         │

             │ build/main/bench.out │           bench.out           │
             │      allocs/op       │  allocs/op    vs base         │

pkg: github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics
                 │ build/main/bench.out │             bench.out             │
                 │        sec/op        │   sec/op     vs base              │
AggregateSpan-12            886.5n ± 4%   846.5n ± 6%  -4.52% (p=0.026 n=6)

                 │ build/main/bench.out │            bench.out            │
                 │         B/op         │     B/op      vs base           │
¹ all samples are equal

                 │ build/main/bench.out │           bench.out           │
                 │      allocs/op       │ allocs/op   vs base           │
¹ all samples are equal

pkg: github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics
                        │ build/main/bench.out │          bench.out           │
                        │        sec/op        │   sec/op     vs base         │

                        │ build/main/bench.out │           bench.out           │
                        │         B/op         │    B/op     vs base           │
¹ all samples are equal

                        │ build/main/bench.out │           bench.out           │
                        │      allocs/op       │ allocs/op   vs base           │
¹ all samples are equal

pkg: github.com/elastic/apm-server/x-pack/apm-server/sampling
               │ build/main/bench.out │             bench.out              │
               │        sec/op        │    sec/op     vs base              │
geomean                  423.0n         438.5n        +3.66%

               │ build/main/bench.out │              bench.out               │
               │         B/op         │     B/op      vs base                │
geomean                    381.8          382.6       +0.21%
¹ all samples are equal

               │ build/main/bench.out │             bench.out              │
               │      allocs/op       │ allocs/op   vs base                │
geomean                    3.742        3.742       +0.00%
¹ all samples are equal

pkg: github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage
                                             │ build/main/bench.out │              bench.out               │
                                             │        sec/op        │    sec/op     vs base                │
WriteTransaction/nop_codec_big_tx-12                  268.4n ± 216%   620.3n ± 54%  +131.15% (p=0.041 n=6)
ReadEvents/proto_codec/1_events-12                    2.040µ ±  15%   1.758µ ±  8%   -13.85% (p=0.015 n=6)
geomean                                               12.80µ          13.07µ          +2.15%

                                             │ build/main/bench.out │               bench.out                │
                                             │         B/op         │     B/op       vs base                 │
ShardedWriteTransactionContended-12                     687.5 ± 26%     948.5 ± 26%  +37.96% (p=0.015 n=6)
geomean                                               11.32Ki         11.42Ki         +0.89%
¹ all samples are equal

                                             │ build/main/bench.out │              bench.out              │
                                             │      allocs/op       │  allocs/op   vs base                │
geomean                                                  123.6         123.6       +0.00%
¹ all samples are equal

```

_report generated with https://pkg.go.dev/golang.org/x/perf/cmd/benchstat_
<!--COMMENT_GENERATED_WITH_ID_bench.diff-->

LikeTheSalad commented:
In this Kibana PR we're adding the android map uploading endpoint which will end up adding map files into ES: https://github.com/elastic/kibana/pull/161252

LikeTheSalad commented on internal/beater/otlp/stacktrace_processor.go:
```diff
@@ -0,0 +1,64 @@
+		if isMobileCrash(event) {
+			err := p.deobfuscate(ctx, event)
+			if err != nil {
+				return err
```
I'm not sure if raising an error in case one event's deobfuscation fails is the recommended way to address this case, given that any unchecked events (if any) will not get deobfuscated because of it. I'm open to suggestions!

LikeTheSalad commented on internal/beater/server.go:
```diff
@@ -191,13 +192,19 @@ func newServer(args ServerParams, listener net.Listener) (server, error) {
+		return server{}, err
+	}
+	mapFetcher := r8.NewMapFetcher(esClient)
+	size := 128
```
I'm not sure what's the format of this value, though I'd like to mention that an Android map could easily get to over 10MB in size, so I wanted to mention this here in case we should provide a higher number in here.

carsonip replied:
I believe this is `size` as in number of entries, not number of bytes of the data cached.

LikeTheSalad commented on internal/beater/server.go:
```diff
@@ -191,13 +192,19 @@ func newServer(args ServerParams, listener net.Listener) (server, error) {
-	otlpBatchProcessor := args.BatchProcessor
+	var otlpBatchProcessor modelprocessor.Chained
+
+	esClient, err := args.NewElasticsearchClient(elasticsearch.DefaultConfig())
```
I couldn't find how the esClient instance is initialized for RUM sourcemaps. I found about this `DefaultConfig` and it allowed me to compile so that's why I used it, although I'm not sure if this is the proper way to initialize it, or if there's another way for us to get an instance of esClient without having to initialize it in here. I'm open to any advice on this!

kruskall replied:
It happens here: https://github.com/elastic/apm-server/blob/main/internal/beater/beater.go#L340-L353

We're passing a callback to the method which creates the fetcher.

LikeTheSalad replied:
Thanks for pointing me in the right direction. I just tried to replicate the same setup for the esClient instance as it's done in `beater.go`, however, I noticed that there are a couple of intricate setup parameters and processes that require creating tracers and callbacks and launching goroutines that seem too many things that I can do wrongly, so I then noticed that this code in `server.go` that adds the OTLP processors is called from `beater.go` from the same place that you mentioned, so I decided to pass the esClient initialization function and config from `beater.go` instead and that way not having to set it all up again in `server.go`. I'm unsure if I'm breaking any rules by doing so, so please have another look and let me know if there's something I need to change with that approach.

LikeTheSalad commented:
Hi @carsonip, @kruskall - I've just added all the changes I was planning to add to provide support for Android stacktrace deobfuscation. In the end, my goal is to add the deobfuscation processing to the OTLP event ingestion where we'd get the android map files from ES and will also cache them to reuse those later. The only thing that is present in the sourcemap RUM processor that I didn't add into this Android one was a listener to invalidate the cache, and the reason is that I couldn't wrap my head around how that works 😔 I'm not sure if that's crucial when it comes to storing things in cache, if so, I'm going to need some help to figure it out.

I've also added some concerns of mine across this PR, so please have a look at those as well, I'm open to any suggestions!

kruskall commented on internal/r8/map_fetcher.go:
```diff
@@ -0,0 +1,100 @@
+
+	// handle error response
+	if resp.StatusCode >= http.StatusMultipleChoices {
+		defer resp.Body.Close()
```
**issue**: we should move this before the if, right after the error check

LikeTheSalad replied:
Thank you, I've just applied the changes.

kruskall commented:
@LikeTheSalad This is great! Thanks for working on this!

I think we should add a system test at this point.

LikeTheSalad commented on internal/beater/server.go:
```diff
@@ -191,13 +192,19 @@ func newServer(args ServerParams, listener net.Listener) (server, error) {
-			modelpb.ProcessBatchFunc(otlp.SetClientMetadata),
-			otlpBatchProcessor,
-		}
+		otlpBatchProcessor = append(otlpBatchProcessor, modelpb.ProcessBatchFunc(otlp.SetClientMetadata))
```
No reason in particular, I've just made the changes to add the stacktrace deobfuscator processor after the client metadata one.

LikeTheSalad commented:
Thank you @kruskall - I've just addressed your comments and I hope it's now better. I'm not familiarized with what you mentioned about a "system test" but if you can give me some hints about how to create it or where to find some examples, I can give it a try.

carsonip commented on internal/beater/server.go:
```diff
@@ -191,14 +199,20 @@ func newServer(args ServerParams, listener net.Listener) (server, error) {
 	}
 
-	otlpBatchProcessor := args.BatchProcessor
+	var otlpBatchProcessor modelprocessor.Chained
```
`args.BatchProcessor` is now ignored, is that intended? I think this will cause everything to send to nowhere.

LikeTheSalad replied:
This is a great point, I didn't notice it 😨 - I'll add it back.

LikeTheSalad replied:
I've just re-added it

carsonip commented on internal/r8/map_fetcher.go:
```diff
@@ -0,0 +1,100 @@
+
+const mapIndex = ".apm-android-map"
+
+var errFetcherUnvailable = errors.New("fetcher unavailable")
```
```suggestion
var errFetcherUnavailable = errors.New("fetcher unavailable")
```

LikeTheSalad replied:
Than you, I'll add the changes locally to get the IDE to replace the name where this is referenced.

carsonip commented on internal/r8/map_fetcher.go:
```diff
@@ -0,0 +1,100 @@
+
+// NewMapFetcher returns a MapFetcher.
+func NewMapFetcher(c *elasticsearch.Client) *MapFetcher {
+	logger := logp.NewLogger(logs.Sourcemap)
```
Create a new `logs.R8` logger name

LikeTheSalad replied:
Thanks, it's there now.

carsonip commented on internal/r8/map_fetcher.go:
```diff
@@ -0,0 +1,100 @@
+		return nil, fmt.Errorf("ES returned unknown status code: %s", resp.Status())
+	}
+
+	body, err := io.ReadAll(resp.Body)
+
+	if err != nil {
+		return nil, err
+	}
+
+	decodedBody, err := base64.StdEncoding.DecodeString(string(body))
+	if err != nil {
+		return nil, fmt.Errorf("failed to base64 decode string: %w", err)
+	}
```
I'm thinking about creating a streaming base64 decoder and zlib reader from resp.Body so that we avoid having multiple copies of large payload in memory, something like `zlib.NewReader(base64.NewDecoder(base64.StdEncoding, resp.Body))`

LikeTheSalad replied:
Thank you, I've applied this change and it looks way cleaner now.

carsonip commented on internal/beater/otlp/map_caching.go:
```diff
@@ -0,0 +1,71 @@
+
+// Fetch fetches an android source map from the cache or wrapped backend.
+func (f *MapCachingFetcher) Fetch(ctx context.Context, name, version string) ([]byte, error) {
+	key := sourcemap.Identifier{
```
I'd argue that we don't need to reuse anything from sourcemap package to avoid an unnecessary dependency on a mostly unrelated package. This package can have its own struct representing an identifier.

LikeTheSalad replied:
Thanks, I've created an ide only for android maps caching.

carsonip commented on internal/beater/otlp/map_caching.go:
```diff
@@ -0,0 +1,71 @@
+package otlp
```
Should this file go into r8 package instead? Since this is more about r8 instead of about otlp.

LikeTheSalad replied:
Makes sense, I've moved it to R8.

carsonip commented on internal/beater/otlp/map_caching.go:
```diff
@@ -0,0 +1,71 @@
+) (*MapCachingFetcher, error) {
+	logger := logp.NewLogger(logs.Sourcemap)
+
+	lruCache, err := lru.NewWithEvict(cacheSize, func(key, value interface{}) {
```
a little concerned about using an in-memory lru cache if the mapping files are going to be >10MB. We'll need to think about how much memory usage this will add. Otherwise, we'll have to cache some on local disk instead.

LikeTheSalad replied:
Sure, what do you suggest we could do to verify whether this will use too much memory or not? Or should we just go ahead and cache it on the local disk right away? I'm not sure how to proceed here.

carsonip replied:
I don't have the background knowledge here to make a decision on this. How large can a mapping file be, in normal cases and extreme cases? We'll have to ensure that `lru_size * mapping_file_size` is within reasonable memory usage while also reserving enough memory for normal APM server operation (e.g. aggregations). I'm highlighting this to make sure that we don't OOM just because we are caching large mapping files, especially for small APM servers (e.g. 1-2GB)

LikeTheSalad replied:
Got it. Our opbeans app's map file size is about 7 MB, so considering that it's a dummy app, I'd say that it's quite easy for an app's map file size to get bigger than 10 MB. However, I don't think it's safe to say that it'd reach sizes bigger than 50 MB, so probably we can say that an extreme case could be around 50 MB (giving it some margin of error just to be safe). On top of that, it is expected that a single application can have multiple versions running at once, and each version must have its own map file. Therefore, I think it's very likely that a lot of map files can be in-memory at a time for a single app in case there are several versions of it being used (which tends to be a common scenario unless the app has some sort of force-upgrade mechanism built-in). I hope it helps to get a better idea of how much memory this feature can take.

carsonip replied:
@simitt what do you think about this matter? Personally at least for smaller instances, the idea of caching a few 50MB mapping files sounds scary to me since it takes away a lot of memory headroom for apm-server. My feeling is that a local disk cache would make sense in this case.

carsonip commented on internal/beater/otlp/stacktrace_processor.go:
```diff
@@ -0,0 +1,64 @@
+}
+
+// NewStacktraceProcessor returns a StacktraceProcessor.
+func NewStacktraceProcessor(fetcher MapFetcher) *StacktraceProcessor {
+	return &StacktraceProcessor{fetcher, crashDeobfuscatorFunc(r8.Deobfuscate)}
```
May be good to not bake in r8 here and have the caller compose it instead. Removes package dependency to r8, at least in this file.
```suggestion
func NewStacktraceProcessor(fetcher MapFetcher, deobfuscator CrashDeobfuscator) *StacktraceProcessor {
	return &StacktraceProcessor{fetcher, deobfuscator}
```

carsonip reviewed (commented):
First pass. Looking good. We'll also need some system tests on this as I suspect the current deobfuscated events are sent to nowhere.

simitt reviewed (changes requested):
Thank you @LikeTheSalad for all your efforts on this. After reviewing the logic in more detail, I unfortunately have several concerns with this direction.

My biggest concern is on an architectural level. Caching potentially large files completely in memory for an ever growing number of different versions (assuming that we would need to support many old versions for mobile) does not seem to be the best path forward. This quickly grows to a substantial RAM requirement, potentially requiring switching to more optimised hardware. Before moving forward in such a direction, we need to discuss alternatives. I just synced with @AlexanderWert on this topic and he'll arrange for you to discuss in further details. But one idea that Alex brought up was to look into de-obfuscating at query time. 

I also have some concerns more specific to implementation, that I would want to be tackled before merging this. But given the doubts on an architectural level, I'll safe that for now and come back to it if we decide to move forward with the de-obfuscation in the apm-server eventually.

simitt commented:
Closing this PR for now; we can re-open if relevant, when the final direction is clarified.
